File: javaScript/blink_io.js
// Program: blink_io.js
// Description: blink any LED connected to ports
//              0 to 13
// Author: Aldo Nunez

// mraa module
var m = require ( 'mraa' );

// errrors  message
message1 = "<port>: 0 | 1 | 2 | 3 | ... | 13";

// number of digital output ports
const PORTS = 14

// states
const STATES = {OFF: 0, ON: 1 };

// time in milliseconds
const T = 100;

if ( process.argv.length < 3 )
{
    console.log ( "Usage: node " + process.argv [ 1 ].split ( "/") [ 5 ] + " <port>" );
    console.log ( message1 );

    process.exit ( 1 );
}

var port = Number ( process.argv [ 2 ] );

// Check if the argument is a number
if ( !Number.isInteger ( port ) )
{
    console.log ( message1 );

    process.exit ( 1 );
}

// Check if the number is between: 0 - 13
if ( ( port < 0  ) || ( port  > PORTS - 1 )  )
{
	console.log ( "<port> must be between: 0..13");
    console.log ( message1 );
    process.exit ( 1 );
}

// choose pin
var led = new m.Gpio ( port );

// set the gpio direction
var response = led.dir ( m.DIR_OUT );
if ( response )
{
    process.exit ( 1 );
}

console.log ( "MRAA Version: " + m.getVersion () );
console.log ( "Platform: " + m.getPlatformName () );
console.log ( "blinking port " + port +  " ..." );
console.log ( "To finish press: Ctrl + c" );

var flag = true;

// set time interval
setInterval ( toogle, T );

// turn on and turn off led
function toogle ()
{
    flag ? led.write ( STATES.ON ): led.write ( STATES.OFF );
    flag ^= true;
}

// catch Ctrl + c
process.on('SIGINT', function()
{
    led.write ( STATES.OFF )
    console.log ( "\nprogram is closing ..." );
    process.exit ( 1 );
});

File: python/blink_io.py
'''
    Program: blink_io.py
    Description: blink any LED connected to pines
                IO0 to IO13
    Author: Aldo Nunez
'''

import mraa as m
import time
import sys


message = "<port>: 0 | 1 | 2 | 3 |...| 13 "
# STATES
OFF = 0
ON = 1

# Number if digital output ports
N = 14

ports = [ i for i in range ( N ) ]

# time
T =  0.1    # 0.1 sec

def main ( argv ):
    try:
        port =  int ( sys.argv [ 1 ] )
    except:
        print ( "Usage: python %s <port>" % sys.argv [ 0 ] )
        print ( message )
        sys.exit ()

    if port < min( ports ) or port > max ( ports ):
        print ( "<port> must be between 0..13" )
        print ( message )
        sys.exit ()

    led = m.Gpio ( port )
    if led == None:
        sys.exit ()

    response = led.dir ( m.DIR_OUT )
    if response != m.SUCCESS:
        m.printError ( response )
        sys.exit()

    print ( "MRAA Version: %s" %  m.getVersion () )
    print ( "Platform: %s" % m.getPlatformName () )
    print ( "blinking port %d ..." %  port )
    print ( "To finish press: Ctrl + c" )

    try:
        while True:
            response = led.write ( ON )
            if response != m.SUCCESS:
                m.printError ( response )

            time.sleep ( T )

            response = led.write ( OFF )
            if response != m.SUCCESS:
                m.printError ( response )

            time.sleep ( T )
    except KeyboardInterrupt:
        led.write ( OFF )
        print ( "\nprogram is closing..." )
        sys.exit ( 0 )

if __name__ == "__main__":
    main ( sys.argv [ 1: ] )
